scripts/o3de/tests/ExportScripts/test_export_source_built_project.py

```diff
@@ -0,0 +1,1296 @@
+
+                    cannot_build_monolithic = (use_sdk and not has_monolithic and use_monolithic)
+
+                    if cannot_build_monolithic:
+                        mock_build_game_targets.assert_not_called()
+                    else:
+                        if check_launcher_type == 0:
+                            mock_build_game_targets.assert_not_called()
+                        else:
+                            mock_build_game_targets.assert_called_once_with(ctx=mock_ctx,
+                                                                            build_config="release",
+                                                                            game_build_path=selected_launcher_build_path,
+                                                                            engine_centric=is_engine_centric,
+                                                                            launcher_types=check_launcher_type,
+                                                                            allow_registry_overrides=False,
+                                                                            tool_config=toolconf,
+                                                                            monolithic_build=use_monolithic,
+                                                                            logger=mock_logger)
+                    mock_build_game_targets.reset_mock()
+
+@pytest.mark.parametrize("is_engine_centric, use_sdk, has_monolithic, use_monolithic", [
+    pytest.param(False, True, True, True),
+    pytest.param(False, True, True, False),
+    pytest.param(False, True, False, True),
+    pytest.param(False, True, False, False),
+    pytest.param(False, False, True, True),
+    pytest.param(False, False, True, False),
+    pytest.param(False, False, False, True),
+    pytest.param(False, False, False, False),
+
+    pytest.param(True, True, True, True),
+    pytest.param(True, True, True, False),
+    pytest.param(True, True, False, True),
+    pytest.param(True, True, False, False),
+    pytest.param(True, False, True, True),
+    pytest.param(True, False, True, False),
+    pytest.param(True, False, False, True),
+    pytest.param(True, False, False, False),
+])
+def test_setup_launcher_layout_directory(tmp_path, is_engine_centric, use_sdk, has_monolithic, use_monolithic):
+    setup_dir_cache = []
+    def cache_params(**kwargs):
+        nonlocal setup_dir_cache
+        setup_dir_cache.append(kwargs)
+    
+    def reset_cache():
+        nonlocal setup_dir_cache
+        setup_dir_cache = []
+
+    test_project_name = "TestProject"
+    test_project_path = tmp_path / "project"
+    test_engine_path = tmp_path / "engine"
+
+    test_o3de_base_path = test_project_path if not is_engine_centric else test_engine_path
+    test_relative_base_path = pathlib.PurePath('this/is/relative')
+    test_absolute_base_path  = tmp_path / "other"
+
```

Comment:
If you want, you can factor this section out into a local function to reduce the line count. This block looks repeated in most of the test cases in this file.

Reply:
yes that would be ideal. The only question I have is how to go about that, since each test case is a separate function.

 If i understand correctly, local functions are only relevant to the function scope they're defined in, and any variable bindings inside the local function are forgotten once left.

Reply:
i meant module scope, not function scope. and the function can return a tuple since this is python.